### PR TITLE
clean pad argument

### PR DESCRIPTION
**Pull request by ilyesarf**
Now, you can do 
```
x.pad((1,1,1,1))
```

### CODE REVIEW

**Comment by geohot** [comment]
So this definitely doesn't belong in mlops, it belongs in Tensor. And it's a pretty complex change, you have to refactor the padding in conv also. But the general idea of a number expanding to a tuple I like.

**Comment by ilyesarf** [comment]
Moved it to Tensor. I also tested conv with the new padding, no problem whatsoever. I don't think that the change is that big. It's just a simple hack that converts a tuple of ints into a tuple of tuples which is how pad in mlops works in the first place.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -204,7 +204,9 @@ def reshape(self, shape, *args) -> Tensor:
   def expand(self, shape, *args) -> Tensor: return mlops.Expand.apply(self, shape=tuple(x if x != -1 else s for s,x in zip(self.shape, argfix(shape, *args))))
   def permute(self, order, *args) -> Tensor: return mlops.Permute.apply(self, order=argfix(order, *args))
   def flip(self, axis, *args) -> Tensor: return mlops.Flip.apply(self, axis=[x if x >= 0 else x+len(self.shape) for x in argfix(axis, *args)])
-  def pad(self, arg:Tuple[Tuple[int, int], ...]) -> Tensor: return mlops.Pad.apply(self, arg=arg) if any(x != (0,0) for x in arg) else self
+  def pad(self, arg:Tuple[Tuple[int, int], ...]) -> Tensor:
```
Comment (geohot) [inline tinygrad/tensor.py:207]: I mean, obviously this type signature is wrong now, right?

What if only some are tuples?

What can you remove from conv?

Comment (ilyesarf) [reply tinygrad/tensor.py]: Fixed edge cases, refactored pad2d and conv.

**Comment by geohot** [comment]
Tons of lines in core code and no tests.

**Comment by jla524** [comment]
These tests are failing
```
FAILED test/test_ops.py::TestOps::test_pad2d - AssertionError: shape mismatch (tinygrad)(3, 3, 3, 10) != (torch)(3, 3, 10, 6)
FAILED test/test_ops.py::TestOps::test_padded_conv2d - AssertionError: shape mismatch (tinygrad)(4, 4, 9, 13) != (torch)(4, 4, 13, 30)
FAILED test/test_ops.py::TestOps::test_padded_conv2d_bs1 - AssertionError: shape mismatch (tinygrad)(1, 4, 9, 11) != (torch)(1, 4, 11, 28)
FAILED test/test_ops.py::TestOps::test_padded_conv3d - AssertionError: shape mismatch (tinygrad)(1, 4, 7, 7, 9) != (torch)(1, 4, 9, 9, 9)
FAILED test/test_ops.py::TestOps::test_padded_conv_transpose2d - AssertionError: shape mismatch (tinygrad)(2, 4, 7, 9) != (torch)(2, 4, 9, 9)
```

**Comment by ilyesarf** [comment]
tests should pass now

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                        19               6
------------------------------------------------------------
lines added in the tinygrad folder: 13
```

**Comment by ilyesarf** [comment]
new tests for pad2d designed by @jla524

**Comment by geohot** [comment]
This is really hard to read. If you want, resubmit with a *very* clear explanation of what this allows.
